### PR TITLE
CSHARP-4477: Implement embeddedDocuments operator in Atlas Search

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: src/MongoDB.Driver/Search/SearchDefinitionRenderContext.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using MongoDB.Bson.Serialization;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Search
+{
+    /// <summary>
+    /// Encapsulates classes needed for rendering Search definitions.
+    /// </summary>
+    /// <typeparam name="TDocument">The type of the document.</typeparam>
+    public sealed class SearchDefinitionRenderContext<TDocument>
```
Comment (sanych-sun) [inline src/MongoDB.Driver/Search/SearchDefinitionRenderContext.cs:10]: Do you think we can implement SearchDefinitionRenderContext as a record with init-only properties and then utilize non-destructive mutation approach by using `with` statement to create a shallow copy of the context?

File: tests/MongoDB.Driver.Tests/Search/SearchDefinitionBuilderTests.cs [hunk 2]
```diff
@@ -187,6 +187,71 @@ public void Compound_typed()
...
+
+            // Compound
+            var compound = subject
+                .Compound()
+                .Must(subject.Text("y", "1"))
+                .Should(subject.Exists("z"));
+
+            AssertRendered(
+                subject.EmbeddedDocument<BsonDocument>("x", compound),
+                "{ embeddedDocument: { path : 'x', operator : { compound : { must : [{ text : { query : '1', path : 'x.y' } }], should : [{ exists : { path : 'x.z' } }] } }}}");
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Search/SearchDefinitionBuilderTests.cs:203]: Minor: missed spaces between curly brackets (here and some below)

File: tests/MongoDB.Driver.Tests/Search/SearchPathDefinitionBuilderTests.cs [hunk 3]
```diff
@@ -33,6 +33,10 @@ public void Analyzer()
             AssertRendered(
                 subject.Analyzer("x", "english"),
                 "{ value: 'x', multi: 'english' }");
+            AssertRendered(
+                subject.Analyzer("x", "english"),
+                "{ value: 'basepath.x', multi: 'english' }",
+                "basepath");
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Search/SearchPathDefinitionBuilderTests.cs:39]: Do we have any tests for more complex basepath? Like "root.nested".

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Tests/Search/SearchPathDefinitionBuilderTests.cs (hunk 3 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Search/SearchPathDefinitionBuilderTests.cs:39]: Good point, added in SearchDefinitionBuilderTests and SearchPathDefinitionBuilderTests

File: src/MongoDB.Driver/Search/SearchDefinition.cs [hunk 4]
```diff
@@ -26,12 +25,13 @@ namespace MongoDB.Driver.Search
...
         /// </summary>
-        /// <param name="documentSerializer">The document serializer.</param>
-        /// <param name="serializerRegistry">The serializer registry.</param>
-        /// <returns>A <see cref="BsonDocument"/>.</returns>
-        public abstract BsonDocument Render(IBsonSerializer<TDocument> documentSerializer, IBsonSerializerRegistry serializerRegistry);
+        /// <param name="renderContext">The render context.</param>
+        /// <returns>
+        /// A <see cref="BsonDocument" />.
+        /// </returns>
+        public abstract BsonDocument Render(SearchDefinitionRenderContext<TDocument> renderContext);
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver/Search/SearchDefinition.cs:34]: Yes.

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**